### PR TITLE
Optimize URLFilter and add option to disable integrated wordlists

### PR DESCRIPTION
**Pull request by its5Q**
This pull request makes two changes to URLFilter:
1. Replacing the naive subword search with a faster Aho-Corasick algorithm for multiple subwords.
2. Adding an argument that allows users to disable the usage of integrated block lists so they can supply their own lists instead.

### CODE REVIEW

**Review by guipenedo — CHANGES_REQUESTED** [review]
Great changes, indeed we should have added a way to disable the default lists before. The perf improvement with ahocorasick is also nice

File: src/datatrove/pipeline/filters/url_filter.py [hunk 1]
```diff
@@ -53,22 +53,32 @@ def __init__(
...
         from tldextract import TLDExtract
 
         super().__init__(exclusion_writer)
         self.soft_word_threshold = soft_word_threshold
-        self.block_listed_domains = extra_domains
-        self.block_listed_url = extra_urls
-        self.banned_words = banned_words
-        self.banned_subwords = banned_subwords
-        self.soft_banned_words = soft_banned_words
+        self.block_listed_domains = parse_list(extra_domains, do_normalize=False) if extra_domains else set()
```
[the remark this answers is not in the document]
Comment (its5Q) [reply src/datatrove/pipeline/filters/url_filter.py:64]: Yeah, I think it's now safe to remove as it's guaranteed to be a set.

**Comment by its5Q** [comment]
Also added a check if the block list is empty, otherwise it would fail with empty block list because the aho-corasick tree isn't built.

File: src/datatrove/pipeline/filters/url_filter.py [hunk 2]
```diff
@@ -27,7 +27,7 @@ def parse_list(line, do_normalize=True):
 
 def get_list(abs_path: str, file_name: str, extra: set = None, do_normalize: bool = True):
```
Comment (guipenedo) [inline src/datatrove/pipeline/filters/url_filter.py:28]:
```suggestion
def get_list(abs_path: str, file_name: str, extra: set, do_normalize: bool = True):
```

File: src/datatrove/pipeline/filters/url_filter.py [hunk 3]
```diff
@@ -53,22 +53,32 @@ def __init__(
...
+        self.block_listed_url = parse_list(extra_urls, do_normalize=False) if extra_urls else set()
+        self.banned_words = parse_list(banned_words) if banned_words else set()
+        self.banned_subwords = parse_list(banned_subwords) if banned_subwords else set()
+        self.soft_banned_words = parse_list(soft_banned_words) if soft_banned_words else set()
+        self.use_integrated_lists = use_integrated_lists
         self._downloaded = False
         self.tldextractor = TLDExtract()
 
+        self.banned_subwords_automaton = ahocorasick.Automaton(ahocorasick.STORE_INTS)
+        for word in self.banned_subwords:
```
Comment (guipenedo) [inline src/datatrove/pipeline/filters/url_filter.py:74]: this probably isn't an issue but if `use_integrated_lists=True`, won't the extra `banned_subwords` be added twice? Once here and again inside `download_data`

Comment (its5Q) [reply src/datatrove/pipeline/filters/url_filter.py:74]: that's probably fine, if the trie already exists add_word should just return False

**Review by guipenedo — APPROVED** [review]

**Comment by guipenedo** [comment]
Ty :)

**Pull request merged**